### PR TITLE
checkpoint based raft snapshot implementation

### PR DESCRIPTION
**Pull request by maxkozlovsky**
Implement raft snapshots as rocksdb checkpoints.

Store raft last applied log id and membership in rocksdb.

Separate state machine state db and log storage db.

Use full_snapshot api for  snapshot transfer.

### CODE REVIEW

File: src/state/network.rs [hunk 1]
```diff
@@ -205,51 +221,129 @@ impl RaftNetwork<TypeConfig> for NetworkConnection {
...
+                        e,
+                    )
+                })?;
+                if n == 0 {
+                    break;
+                }
+                let req = SnapshotFileChunkRequest {
+                    snapshot_id: snapshot.meta.snapshot_id.clone(),
+
+                    vote: serde_json::to_string(&vote).unwrap(),
```
Comment (diptanu) [inline src/state/network.rs:283]: Remove unwrap

File: src/state/network.rs [hunk 2]
```diff
@@ -205,51 +221,129 @@ impl RaftNetwork<TypeConfig> for NetworkConnection {
...
+                    name: entry.file_name().into_string().unwrap().to_string(),
```
Comment (diptanu) [inline src/state/network.rs:284]: same

File: src/state/network.rs [hunk 3]
```diff
@@ -205,51 +221,129 @@ impl RaftNetwork<TypeConfig> for NetworkConnection {
...
+                    data: buf,
```
Comment (diptanu) [inline src/state/network.rs:285]: Can we record how many bytes were sent as metric?

File: src/state/store/mod.rs [hunk 4]
```diff
@@ -140,165 +152,298 @@ pub struct StoredSnapshot {
...
+    async fn new(db_path: PathBuf) -> Result<StateMachineStore, StorageError<NodeId>> {
+        let (tx, rx) = tokio::sync::watch::channel(StateChangeId::new(std::u64::MAX));
         let (gc_tasks_tx, _) = broadcast::channel(100);
+
+        delete_incomplete_snapshots(&db_path).map_err(|e| {
+            StorageError::from_io_error(ErrorSubject::Snapshot(None), ErrorVerb::Delete, e)
+        })?;
+
+        let mut snapshots = read_snapshots(&db_path)?;
+        println!("snapshots: {:?}", snapshots);
```
Comment (diptanu) [inline src/state/store/mod.rs:272]: Remove print

File: src/state/store/mod.rs [hunk 5]
```diff
@@ -140,165 +152,298 @@ pub struct StoredSnapshot {
...
+            snapshots.push(Snapshot {
+                meta: Default::default(),
+                snapshot: Box::new(SnapshotData {
+                    snapshot_dir: db_path,
+                }),
+            });
+        }
+
+        // live db is the directory with the latest applied log id
+        let live_snapshot = snapshots.pop().unwrap();
```
Comment (diptanu) [inline src/state/store/mod.rs:296]: Remove unwrap

File: src/state/store/mod.rs [hunk 6]
```diff
@@ -140,165 +152,298 @@ pub struct StoredSnapshot {
...
+
+        // Remove all but last snapshot directories
+        let snapshot = if snapshots.len() >= 1 {
+            for snapshot in snapshots[..snapshots.len() - 1].iter() {
+                fs::remove_dir_all(&snapshot.snapshot.snapshot_dir).map_err(|e| {
+                    StorageIOError::read_snapshot(None, anyhow!("Failed to remove snapshot: {}", e))
+                })?;
+            }
+            Some(snapshots.pop().unwrap())
```
Comment (diptanu) [inline src/state/store/mod.rs:305]: same

**Review by diptanu — CHANGES_REQUESTED** [review]

File: src/state/network.rs (hunk 1 above)
Comment (maxkozlovsky) [reply src/state/network.rs]: ok

File: src/state/network.rs (hunk 2 above)
Comment (maxkozlovsky) [reply src/state/network.rs]: ok

File: src/state/network.rs (hunk 3 above)
Comment (maxkozlovsky) [reply src/state/network.rs:292]: ok

File: src/state/store/mod.rs (hunk 4 above)
Comment (maxkozlovsky) [reply src/state/store/mod.rs]: ok

**Pull request merged**